gradle-guide-error-prone/src/test/java/com/palantir/gradle/guide/errorprone/NonAbstractGradleTypeFieldsTest.java

```diff
@@ -0,0 +1,398 @@
+                """;
+
+        @Test
+        void provider_field_in_constructor_should_fail() {
+            compilationTestHelper
+                    .addSourceLines(
+                            "FooExtension.java",
+                            "import org.gradle.api.provider.Property;",
+                            "import org.gradle.api.Project;",
+                            "abstract class FooExtension {",
+                            "    private final Property<String> property;",
+                            "    public FooExtension(Project project) {",
+                            "        this.property = project.getObjects().property(String.class);",
+                            "    }",
+                            "}")
+                    .addSourceLines(
+                            "FooPlugin.java",
+                            """
+                            import org.gradle.api.Plugin;
+                            import org.gradle.api.Project;
+                            public class FooPlugin implements Plugin<Project> {
+                                @Override
+                                public void apply(Project project) {
+                                    // BUG: Diagnostic contains: declare class fields directly on Tasks or Extensions
+                                    project.getExtensions().create("foo", FooExtension.class);
+                                }
+                            }
+                            """)
+                    .doTest();
+        }
+
+        @Test
+        void file_collection_field_in_constructor_should_fail() {
+            compilationTestHelper
+                    .addSourceLines(
+                            "FooExtension.java",
+                            "import org.gradle.api.file.FileCollection;",
+                            "import org.gradle.api.Project;",
+                            "abstract class FooExtension {",
+                            "    private final FileCollection files;",
+                            "    public FooExtension(Project project) {",
+                            "        this.files = project.files(\"somefile.txt\");",
+                            "    }",
+                            "}")
+                    .addSourceLines(
+                            "FooPlugin.java",
+                            """
+                            import org.gradle.api.Plugin;
+                            import org.gradle.api.Project;
+                            public class FooPlugin implements Plugin<Project> {
+                                @Override
+                                public void apply(Project project) {
+                                    // BUG: Diagnostic contains: declare class fields directly on Tasks or Extensions
+                                    project.getExtensions().create("foo", FooExtension.class);
+                                }
+                            }
+                            """)
```

Comment:
Extract this out to a variable to avoid duplication

Reply:
Even better, make a new method to run this test inside this nested test class, you can then add `@Language("Java")` to get syntax highlighting on the extension code.